### PR TITLE
Fix comparable in switch

### LINKED ISSUE #10892: string & Type intersection produce error in switch statement
**TypeScript Version:**  2.0.2

**Code**

``` ts
class MyBrand
{
    private _a;
}

function test(strInput: string & MyBrand, numInput: number & MyBrand) {
    switch(strInput)
    {
        case "a": // error: Type '"a"' is not comparable to type 'string & MyBrand'
        return 1;
    }

    switch(numInput)
    {
        case 1: // error: Type '1' is not comparable to type 'number & MyBrand'
        return 1;
    }

    return 0;
}
```

**Expected behavior:**
No error reported in case statements.

**Actual behavior:** 
See above error report inline.

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #10892

`switch` should use the same rules as `===`, so I pushed the calls to `getBaseTypeOfLiteralType` into `isTypeEqualityComparableTo`, which both `checkBinaryLikeExpression` and `checkSwitchStatement` call.

### CODE REVIEW

**Comment by sandersn** [comment]
@DanielRosenwasser I found a pretty good solution to make this fix cleaner. It drops more literal types into error messages, but is otherwise not bad.

**Comment by sandersn** [comment]
@DanielRosenwasser I believe you knew about this code since you refactored comparability this year. Can you review it?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -16903,7 +16903,13 @@ namespace ts {
                     // TypeScript 1.0 spec (April 2014): 5.9
                     // In a 'switch' statement, each 'case' expression must be of a type that is comparable
                     // to or from the type of the 'switch' expression.
-                    const caseType = checkExpression(caseClause.expression);
+                    let caseType = checkExpression(caseClause.expression);
+                    const caseIsLiteral = isLiteralType(caseType);
+                    const expressionIsLiteral = isLiteralType(expressionType);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: This can be cached.

**Comment by DanielRosenwasser** [comment]
@aozgaa weren't you working on a similar bug regarding intersection type comparability? What was the outcome of that?

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/checker.ts]: done

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -16903,10 +16904,16 @@ namespace ts {
                     // TypeScript 1.0 spec (April 2014): 5.9
                     // In a 'switch' statement, each 'case' expression must be of a type that is comparable
                     // to or from the type of the 'switch' expression.
-                    const caseType = checkExpression(caseClause.expression);
-                    if (!isTypeEqualityComparableTo(expressionType, caseType)) {
+                    let caseType = checkExpression(caseClause.expression);
+                    const caseIsLiteral = isLiteralType(caseType);
+                    let literalExpressionType = expressionType;
```
Comment (sandersn) [inline src/compiler/checker.ts]: Thanks, that name is better.

**Comment by ahejlsberg** [comment]
👍

**Pull request merged**